File: solutions/ocaml/071-minimum-knight-moves.ml
(*
  Problem 71: Minimum Knight Moves (LeetCode 1197)
  Difficulty: Med
  Language: OCaml
*)
let min_knight_moves x y =
  let tx = abs x and ty = abs y in
  let moves = [(1,2);(2,1);(2,-1);(1,-2);(-1,-2);(-2,-1);(-2,1);(-1,2)] in
  let module S = Set.Make(struct type t = int*int let compare = compare end) in
  let rec bfs queue vis = match queue with
    | [] -> -1
    | (cx,cy,d)::rest ->
      if cx = tx && cy = ty then d
      else
        let nexts = List.filter_map (fun (dx,dy) ->
          let nx = cx+dx and ny = cy+dy in
          if nx >= -2 && ny >= -2 && not (S.mem (nx,ny) vis)
          then Some (nx,ny,d+1) else None) moves in
        let vis' = List.fold_left (fun s (a,b,_) -> S.add (a,b) s) vis nexts in
        bfs (rest @ nexts) vis'
  in bfs [(0,0,0)] (S.singleton (0,0))

File: solutions/haskell/071-minimum-knight-moves.hs
{-
  Problem 71: Minimum Knight Moves (LeetCode 1197)
  Difficulty: Med
  Language: Haskell
-}
import qualified Data.Set as S
minKnightMoves :: Int -> Int -> Int
minKnightMoves x y = bfs [(0,0,0)] S.empty
  where
    tx = abs x; ty = abs y
    moves = [(1,2),(2,1),(2,-1),(1,-2),(-1,-2),(-2,-1),(-2,1),(-1,2)]
    bfs [] _ = -1
    bfs ((cx,cy,d):rest) vis
      | cx == tx && cy == ty = d
      | otherwise =
          let nexts = [(cx+dx,cy+dy,d+1) | (dx,dy) <- moves,
                       cx+dx >= -2, cy+dy >= -2,
                       S.notMember (cx+dx,cy+dy) vis]
              vis' = foldl (\s (a,b,_) -> S.insert (a,b) s) vis nexts
          in bfs (rest ++ nexts) vis'
